File: src/main/java/com/github/bjoernpetersen/deskbot/model/PluginWrapper.java
package com.github.bjoernpetersen.deskbot.model;

import com.github.bjoernpetersen.jmusicbot.IdPlugin;
import com.github.bjoernpetersen.jmusicbot.NamedPlugin;
import com.github.bjoernpetersen.jmusicbot.ProviderManager;
import com.github.bjoernpetersen.jmusicbot.ProviderManager.State;
import com.github.bjoernpetersen.jmusicbot.config.Config;
import java.util.logging.Logger;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javax.annotation.Nonnull;

public class PluginWrapper<P extends IdPlugin> {

  @Nonnull
  private static final Logger log = Logger.getLogger(PluginWrapper.class.getName());

  @Nonnull
  private final ProviderManager manager;
  @Nonnull
  private final Config.BooleanEntry activeConfig;
  @Nonnull
  private P plugin;
  @Nonnull
  private final BooleanProperty active;
  @Nonnull
  private final ObservableList<Config.Entry> entries;

  public PluginWrapper(@Nonnull Config config, @Nonnull ProviderManager providerManager,
      @Nonnull P plugin) {
    this.manager = providerManager;
    this.activeConfig = config.booleanEntry(
        plugin.getClass(),
        "enable",
        "Enables plugin: " + plugin.getReadableName(),
        false
    );

    this.plugin = plugin;
    this.active = new SimpleBooleanProperty(false);
    this.entries = FXCollections.observableArrayList();

    this.manager.addStateListener(getPlugin(), (old, state) -> {
      if (old == State.INACTIVE && state != State.INACTIVE) {
        active.set(true);
      }
    });
    this.active.addListener(((observable, oldValue, newValue) -> {
      activeConfig.set(newValue);
      entries.clear();
      if (newValue) {
        entries.addAll(manager.getConfigEntries(getPlugin()));
      } else {
        manager.destructConfigEntries(getPlugin());
      }
    }));
    this.active.set(activeConfig.get());
  }

  public ObservableList<? extends Config.Entry> getConfigEntries() {
    return this.entries;
  }

  public void setActive(boolean active) {
    activeProperty().set(active);
  }

  public boolean isActive() {
    return activeProperty().get();
  }

  @Nonnull
  public BooleanProperty activeProperty() {
    return active;
  }

  @Nonnull
  public P getPlugin() {
    return this.plugin;
  }
}

File: src/main/java/com/github/bjoernpetersen/deskbot/model/PluginWrapper.kt
package com.github.bjoernpetersen.deskbot.model

import com.github.bjoernpetersen.jmusicbot.*
import com.github.bjoernpetersen.jmusicbot.config.Config
import com.github.bjoernpetersen.jmusicbot.playback.PlaybackFactory
import com.github.bjoernpetersen.jmusicbot.provider.Provider
import com.github.bjoernpetersen.jmusicbot.provider.Suggester
import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.collections.FXCollections
import javafx.collections.ObservableList

interface ObservablePluginWrapper<out P : Plugin> : PluginWrapper<P> {
  val enabledEntry: Config.BooleanEntry

  fun enabledProperty(): BooleanProperty
  var isEnabled: Boolean
    get() = enabledProperty().get()
    set(value) = enabledProperty().set(value)

  val observableConfigEntries: ObservableList<Config.Entry>
}

private fun init(config: Config, wrapper: ObservablePluginWrapper<*>) {
  wrapper.enabledProperty().addListener { _, old, new ->
    if (old != new) {
      if (new) wrapper.initializeConfigEntries(config)
      else wrapper.destructConfigEntries()
      wrapper.enabledEntry.set(new)
    }
  }
  wrapper.addStateListener { old, new ->
    if (old === Plugin.State.INACTIVE && new === Plugin.State.CONFIG) {
      wrapper.observableConfigEntries.addAll(wrapper.configEntries)
    } else if (old === Plugin.State.CONFIG && new === Plugin.State.INACTIVE) {
      wrapper.observableConfigEntries.clear()
    }
  }
  wrapper.isEnabled = wrapper.enabledEntry.value
}

class ObservablePlaybackFactoryWrapper(config: Config, factory: PlaybackFactory) :
    DefaultPlaybackFactoryWrapper(factory), ObservablePluginWrapper<PlaybackFactory> {

  override val enabledEntry: Config.BooleanEntry = config.BooleanEntry(factory.javaClass, "enable", "", false)
  private val enabledProperty: BooleanProperty = SimpleBooleanProperty(false)
  override fun enabledProperty() = enabledProperty

  override val observableConfigEntries: ObservableList<Config.Entry> = FXCollections.observableArrayList()

  init {
    init(config, this)
  }
}

class ObservableProviderWrapper(config: Config, provider: Provider) : DefaultProviderWrapper(provider),
    ObservablePluginWrapper<Provider> {

  override val enabledEntry: Config.BooleanEntry = config.BooleanEntry(provider.javaClass, "enable", "", false)
  private val enabledProperty: BooleanProperty = SimpleBooleanProperty(false)
  override fun enabledProperty() = enabledProperty

  override val observableConfigEntries: ObservableList<Config.Entry> = FXCollections.observableArrayList()

  init {
    init(config, this)
  }
}

class ObservableSuggesterWrapper(config: Config, suggester: Suggester) : DefaultSuggesterWrapper(suggester),
    ObservablePluginWrapper<Suggester> {

  override val enabledEntry: Config.BooleanEntry = config.BooleanEntry(suggester.javaClass, "enable", "", false)
  private val enabledProperty: BooleanProperty = SimpleBooleanProperty(false)
  override fun enabledProperty() = enabledProperty

  override val observableConfigEntries: ObservableList<Config.Entry> = FXCollections.observableArrayList()

  init {
    init(config, this)
  }
}

class ObservableAdminPluginWrapper(config: Config, plugin: AdminPlugin) :
    DefaultAdminPluginWrapper(plugin), ObservablePluginWrapper<AdminPlugin> {

  override val enabledEntry: Config.BooleanEntry = config.BooleanEntry(plugin.javaClass, "enable", "", false)
  private val enabledProperty: BooleanProperty = SimpleBooleanProperty(false)
  override fun enabledProperty() = enabledProperty

  override val observableConfigEntries: ObservableList<Config.Entry> = FXCollections.observableArrayList()

  init {
    init(config, this)
  }
}
